### PR TITLE
Retrieve Qin-Denton and leapseconds from new source (Closes #233)

### LINKED ISSUE #233: OMNI data is only available through January 2017
Updating the OMNI database with [spacepy.toolbox.update()](https://pythonhosted.org/SpacePy/autosummary/spacepy.toolbox.update.html#spacepy.toolbox.update) only updates the database up to January 31, 2017.  This is probably because the source used for OMNI data ([ViRBO](http://virbo.org/)) is no longer maintained.  NASA's [Space Physics Data Facility](https://spdf.gsfc.nasa.gov/) looks like it is the appropriate place to get this data now.

The spacepy [omni documentation](https://pythonhosted.org/SpacePy/omni.html) states that
>"The omni module primarily manages the hourly OMNI2 and Qin-Denton data, which are sourced from the Virtual Radiation Belt Observatory (ViRBO), who maintain these data sources."

However, the [ViRBO](http://virbo.org/) page states 
>"The ViRBO project was discontinued in 2014 and these pages are no longer being updated."

The contents of `~/.spacepy/spacepy.rc` confirm that both the omni2 and the qindenton data is being sourced from the ViRBO site.
```
[spacepy]
  
#apply_plot_styles: True #default in SpacePy 0.1.6
#enable_deprecation_warning: True #default in SpacePy 0.1.6
#leapsec_url: ftp://maia.usno.navy.mil/ser7/tai-utc.dat #default in SpacePy 0.1.6
#ncpus: 8 #default in SpacePy 0.1.6
#omni2_url: http://virbo.org/ftp/OMNI/OMNI2/merged/latest/OMNI_OMNI2-latest.cdf.zip #default in SpacePy 0.1.6
#psddata_url: http://spacepy.lanl.gov/repository/psd_dat.sqlite #default in SpacePy 0.1.6
#qindenton_url: http://virbo.org/ftp/QinDenton/hour/merged/latest/WGhour-latest.d.zip #default in SpacePy 0.1.6
#support_notice: True #default in SpacePy 0.1.6
```
After running [spacepy.toolbox.update()](https://pythonhosted.org/SpacePy/autosummary/spacepy.toolbox.update.html#spacepy.toolbox.update), both `~/.spacepy/data/omni2data.h5` and `~/.spacepy/data/omnidata.h5` are modified, however, the last time stamp in `omnidata.h5` is `2017-01-31T23:00:00` and the last time stamp in `omni2data.h5` is `2012-12-31T23:30:00`. 

omnidataSS
omni2dataSS

Both low and high resolution OMNI data can be found from [SPDF](https://spdf.gsfc.nasa.gov/pub/data/omni/).

### Minimal example to reproduce issue:

```
import spacepy.toolbox as tb
tb.update(omni=True)

import spacepy.omni as om
print('OMNI RANGE:', om.omnirange())

# This produces an error
import spacepy.time as spt
import datetime as dt

st = dt.datetime(2017,5,27)
en = dt.datetime(2017,5,28)
ticks = spt.tickrange(st, en, dt.timedelta(hours=1), 'UTC')
data = om.get_omni(ticks)
```

### Error message/Traceback:

This is the full output from the code above.  Even after updating, the latest date where data is available is still January 1, 2017, which causes and error to be raised when data from May 2017 is requested.

```
Retrieving Qin_Denton file ...
Download Progress ...100%

Now processing (this may take a minute) ...
Now saving... 
OMNI RANGE: (datetime.datetime(1963, 11, 27, 14, 0), datetime.datetime(2017, 1, 31, 23, 0))
Traceback (most recent call last):
  File "update_spacepy.py", line 14, in <module>
    data = om.get_omni(ticks)
  File "/anaconda2/envs/python3.6/lib/python3.6/site-packages/spacepy/omni.py", line 254, in get_omni
    raise ValueError('Requested dates are outside data range')
ValueError: Requested dates are outside data range
```

### OS, Python version, and dependency version information:

**OS:** Darwin-17.7.0-x86_64-i386-64bit
**Python:** 3.6.8
numpy=1.16.2
scipy=1.2.1
matplotlib=3.0.2

### Version of SpacePy

spacepy=0.2.1

### PR DESCRIPTION
**Pull request by jtniehof**
The ViRBO source for the Qin-Denton formatted OMNI files is no longer maintained (#233). This PR adds the daily files from the New Mexico Consortium (used to support Van Allen Probes) to extend up to the present; it is merged with the ViRBO file. Because these are daily files retrieved over HTTPS, and connection setup is expensive, it also adds the ability to reuse HTTP connections between requests. This isn't very simple or robust code, but it's the difference between five minutes to check all the files and thirty minutes. It will always fall back to the previous code if necessary.

This PR also retrieves the leapsecond file from NASA MODIS, as the USNO is still down.

- [X] Pull request has descriptive title
- [X] Pull request gives overview of changes
- [X] New code has inline comments where necessary
- [X] Any new modules, functions or classes have docstrings consistent with SpacePy style
- [X] Added an entry to CHANGELOG if fixing a major bug or providing a major new feature
- [X] New features and bug fixes should have unit tests
- [X] Relevant issues are linked to (e.g. `See issue #` or `Closes #`)

### CODE REVIEW

**Comment by jtniehof** [comment]
Yes, we should discuss long-term offline.

**Comment by balarsen** [comment]
> @balarsen , I'd appreciate it if you can pull and run...you have a tendency to find fragilities and I'd like this one hammered on.

I will try and do that right now. I looked at the code and am happy from a look, let me run it.

**Comment by balarsen** [comment]
Pulled and tested while attached to the work VPN (proxies etc). 
Here is a bit of a running commentary while I poke this, sorry for length and formatting. 

```
Now processing (this may take a minute) ...
Finding Q-D daliy files to download ...
Listing files ...100%
```
This was fast

```
Retrieving Q-D daliy files ...
Download Progress ...1%
```
this is super slow (nearly an hour). Maybe not for this PR but we can bundle them up or do something if that helps. 

```
Reading/processing Q-D daily files ...
Now saving...
Finding OMNI2 files to download ...
Listing files ...25%
```
fast


```
Retrieving OMNI2 files ...
Download Progress ...2%
```
medium speed, (5min). 

Side Note: OMNI is magic :)
```
Reading OMNI2 files ...
Retrieving leapseconds file ...
Download Progress ...107%
```
That's a lot of percents. 

----------------------------
Exited and ran again. 
```
Retrieving Q-D daliy files ...
Download Progress ...2%
```
Still crazy slow on the second run, No indication that it is not grabbing all the files again.

**Comment by balarsen** [comment]
So with `keepalive=True` the `'Falling back to slow'` printed in 
```
Retrieving Qin_Denton file ...
Now processing (this may take a minute) ...
Finding Q-D daliy files to download ...
```
but not in 
`Retrieving Q-D daliy files ...` (I did not let it finish). 

With 
`keepalive=False`
`'Falling back to slow'` did not print and it is the same slow speed. 

So from my watch perspective I see no difference with `keepalive`

**Comment by jtniehof** [comment]
`Falling back to slow` isn't going to print if `keepalive=False` because it's going straight to the slow. It will also _only_ print when doing the "finding"...if the keepalive fails there, it never tries again.

If you've been trying this on a work machine (or with VPN up, which looks like you have), try it on a home connection. Requiring a web proxy will definitely fall back to the slow stuff.

The Q-D non-daily, BTW, should be a fairly obvious difference between the first and subsequent times, since it's a fairly large single file. I forgot to mention: this PR now caches that file, before we deleted it and redownloaded every time, even if it didn't change.

**Comment by balarsen** [comment]
We are rightly loathe to bring in other dependencies but there have to be multiple packages to download stuff that we could just use and not have to do this home-rolling. Not that the others are totally better but at least not our problem.

**Comment by balarsen** [comment]
> Off the VPN
> 
> ```
> In [2]: spacepy.toolbox.update(all=True)
> Retrieving Qin_Denton file ...
> Now processing (this may take a minute) ...
> Finding Q-D daliy files to download ...
> Listing files ...100%
> 
> Retrieving Q-D daliy files ...
> Download Progress ...5%
> ```
> 
> This seems 2-3x faster. The `Falling back to slow` did not print.

This functioned and made me happy with the changes.

**Comment by jtniehof** [comment]
Yes, long term I think we want to answer the question of if we're going to be in the business of downloading data, and pull in an appropriate dependency. 100%.

**Comment by balarsen** [comment]
I am done and happy with this one, @drsteve ?

**Comment by jtniehof** [comment]
Another ping to @drsteve ...anything I can do to help make this reviewable?

**Comment by drsteve** [comment]
Can you make my life and work schedule manageable? That'd help...
Seriously though, I figured this would take me a little longer to work
through than typical reviews, so I just haven't been able to carve out a
chunk of time.
I'll see what I can do about working through things in the next day or two.

**Comment by jtniehof** [comment]
I'll work on the time-turner. This is definitely a big chunk. Thanks.

File: spacepy/toolbox/__init__.py [hunk 1]
```diff
@@ -736,31 +743,43 @@ def dictree(in_dict, verbose=False, spaces=None, levels=True, attrs=False, **kwa
...
+        included.
     """
-    import spacepy.pycdf
-    import spacepy.pycdf.istp
-    if omni2url is None:
-        omni2url = spacepy.config['omni2_url']
+    name = '' if name is None else '{} '.format(name)
     #Find all the files to download
-    print("Finding OMNI2 files to download ...")
+    print("Finding {}files to download ...".format(name))
```
[the remark this answers is not in the document]
Comment (jtniehof) [reply spacepy/toolbox/__init__.py:777]: We don't support 2.6 or 3.0 anymore.

File: spacepy/toolbox/__init__.py [hunk 2]
```diff
@@ -769,25 +788,27 @@ def _get_cdaweb_omni2(omni2url=None):
...
         p.feed(data)
         p.close()
         for f in p.links:
-            if not re.match(r'omni2_h0_mrg1hr_\d{8}_v\d+\.cdf', f):
+            if not re.match(pattern, f):
                 continue
             downloadme[f] = yearurl + f
         progressbar(i + 1, 1, len(yearlist), text='Listing files')
-    print("Retrieving OMNI2 files ...")
+    print("Retrieving {}files ...".format(name))
```
[the remark this answers is not in the document]
Comment (jtniehof) [reply spacepy/toolbox/__init__.py:810]: The very first line of the function adds the space unless there's no name.

File: spacepy/toolbox/__init__.py [hunk 3]
```diff
@@ -796,11 +817,147 @@ def _get_cdaweb_omni2(omni2url=None):
...
+    =======
+    SpaceData
+        The data extracted from the Q-D dataset, fully processed for saving
+        as SpacePy HDF5 OMNI data.
+    """
+    if qd_daily_url is None:
+        qd_daily_url = spacepy.config['qd_daily_url']
+    datadir = os.path.join(spacepy.DOT_FLN, 'data', 'qindenton_daily_files')
+    _crawl_yearly(qd_daily_url, r'QinDenton_\d{8}_hour.txt',
+                  datadir, name='Q-D daliy')
```
[the remark this answers is not in the document]
Comment (jtniehof) [reply spacepy/toolbox/__init__.py]: Fixed in b6ed736

File: spacepy/toolbox/__init__.py [hunk 4]
```diff
@@ -951,28 +1117,68 @@ def get_url(url, outfile=None, reporthook=None, cached=False):
...
-                return None
+                if keepalive:
+                    r.read()
+                else:
+                    r.close()
+                return (None, conn) if keepalive else None
+        if keepalive: # Replace previous header request with full get
+            r.read()
+            conn.request('GET', path, headers=clheaders)
+            r, headers= checkresponse(conn)
```
[the remark this answers is not in the document]
Comment (jtniehof) [reply spacepy/toolbox/__init__.py]: Fixed in b6ed736

**Comment by jtniehof** [comment]
Can you provide more details on what happened with OMNI2? I'm not finding a codepath that would wind up with a corrupted file unless you ran out of disk space while writing--it downloads all the data first, and then writes to file. So unless there's some error condition happening in the read from the sever that isn't resulting in an exception (a problem that would need fixing), I'm not sure how this would happen. At any rate, 4d51c11 adds a "cached" option (default to True) that carries through to the get_url calls. That at least keeps the user from needing to know where the files are to delete.

Better messages are in the commit with typo fixes (b6ed736). There's a problem with a linebreak spacing stuff out improperly, that's a progressbar issue, #350.

I'm going to try to be really precise regarding speed. The very slow thing is setting up an HTTPS connection. This is no faster, and no slower, if you're setting up the connection to get the header information for the check on date, or to download the data. That's why there's so much work going into trying to keep the connection open. Probably the second slowest thing is the actual HTTP requrest. Again, same if it's checking or not. I suspect the actual download is epsilon on most connections, given the size of the files. In the case of downloading blind, there is one HTTP request per file and the data transfer time. If checking, up-to-date files take one HTTP request (saving the transfer time) and not-up-to-date files requre two requests and the transfer time. So it's definitely slower in the not-up-to-date case. Behind a proxy, every place I say "HTTP request" add "and HTTPS connection". The upshot is that, for the case of having all the files up to date, the cached version is six minutes, the uncached is something like six minutes and twenty seconds. This is not a big speedup, but it isn't a slowdown, and if we set up a source with (say) monthly files the speedup should get substantially better.

**Comment by jtniehof** [comment]
@drsteve , I'm not marking resolved on the per-line comments, so can you do so if you're happy with the solution?

File: spacepy/toolbox/__init__.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (drsteve) [reply spacepy/toolbox/__init__.py:777]: I thought the change was later than that, but it looks like I was wrong.

**Comment by drsteve** [comment]
> Can you provide more details on what happened with OMNI2? I'm not finding a codepath that would wind up with a corrupted file unless you ran out of disk space while writing

I honestly don't know how it would've got there myself. It was both of the 1964 CDFs. I had run out of battery mid-update, so my laptop shut down. I was pretty sure those were already downloaded though, and I'm not sure how two files would be corrupted by a shutdown mid-update. Still, I figure if it happened to me it can happen to someone else and having a cleaner path to fixing bad files is appreciated.

> This is not a big speedup, but it isn't a slowdown, and if we set up a source with (say) monthly files the speedup should get substantially better.

On this, could we not just have NMC also serve tarballs? Even annual tarballs are small/should be quick to download, and it won't require either changing the processing code or doing post-processing to make another set of files. This could easily be continued by whatever "permanent" home is found for calculating/serving the data.

I'll wait for your thoughts, but I think I'm ready to merge if you're okay with where this sits.

**Comment by jtniehof** [comment]
We could do tarballs, although that would involve some further code change. My preference would be to take that up in the context of a permanent home--if the long-term data product is best done as daily files but with annual tarballs, say, that could be pretty reasonable.

I'm good for the merge.

**Pull request merged**